includes/embeds/class-amp-gallery-embed-handler.php

```diff
@@ -16,264 +15,193 @@
-			 * while the <amp-carousel> usually has a width of 600 and a height of 480.
-			 * That often means very low-resolution images.
-			 * So fall back to the normal 'gallery' shortcode callback, gallery_shortcode().
-			 */
-			return '';
+			$this->process_gallery_embed( $is_carousel, $is_lightbox, $node, $img_elements );
 		}
-
-		return $this->shortcode( $attributes );
 	}
 
 	/**
-	 * Render.
+	 * Get the caption element for the specified image element.
 	 *
-	 * @param array $args Args.
-	 * @return string Rendered.
+	 * @param DOMElement $img_element Image element.
+	 * @return DOMElement|null The caption element, otherwise `null` if it could not be found.
 	 */
-	public function render( $args ) {
-		$dom                        = new Document();
-		$this->did_convert_elements = true;
-
-		$args = wp_parse_args(
-			$args,
-			[
-				'images' => false,
-			]
-		);
-
-		if ( empty( $args['images'] ) ) {
-			return '';
+	protected function get_caption_element( DOMElement $img_element ) {
+		$parent_element = $this->get_parent_container_for_image( $img_element );
+
+		if ( ! $parent_element instanceof DOMElement ) {
+			return null;
 		}
 
-		$images = new ElementList();
-		foreach ( $args['images'] as $props ) {
-			$image_atts = [
-				'src'    => $props['url'],
-				'width'  => $props['width'],
-				'height' => $props['height'],
-				'layout' => 'responsive',
-				'alt'    => $props['alt'],
-			];
-			if ( ! empty( $props['srcset'] ) ) {
-				$image_atts['srcset'] = $props['srcset'];
-			}
+		// The caption should be next immediate element located next to the parent container of the image.
+		$caption_element = $parent_element->nextSibling;
 
-			if ( ! empty( $args['lightbox'] ) ) {
-				$image_atts['lightbox'] = '';
-			}
-			$image = AMP_DOM_Utils::create_node(
-				$dom,
-				'img',
-				$image_atts
-			);
+		while (
+			$caption_element
+			&& ( ! $caption_element instanceof DOMElement || ! AMP_DOM_Utils::has_class( $caption_element, 'wp-caption-text' ) )
+		) {
+			$caption_element = $caption_element->nextSibling;
+		}
+
+		if ( $caption_element instanceof DOMElement && 'figcaption' !== $caption_element->nodeName ) {
+			// Transform the caption element into a `figcaption`. This not only allows the `amp-lightbox` to correctly
+			// detect and display the caption, but it is also semantically correct as the parent element will be a `figure`.
+			$figcaption_element = AMP_DOM_Utils::create_node( Document::fromNode( $caption_element ), 'figcaption', [] );
 
-			if ( ! empty( $props['href'] ) ) {
-				$previous_image = $image;
-				$image          = AMP_DOM_Utils::create_node(
-					$dom,
-					'a',
-					[
-						'href' => $props['href'],
-					]
-				);
-				$image->appendChild( $previous_image );
+			while ( $caption_element->hasChildNodes() ) {
+				$child = $caption_element->childNodes->item( 0 );
+				$child = $caption_element->ownerDocument->importNode( $child, true );
+				$figcaption_element->appendChild( $child );
```

Comment:
Why `importNode`? It's already in this document.

```suggestion
			while ( $caption_element->firstChild ) {
				$figcaption_element->appendChild( $caption_element->firstChild );
```

Reply:
Remnant of some old code I had.